File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/sqlite/SQLiteHabitList.java
/*
 * Copyright (C) 2016-2021 Álinson Santos Xavier <git@axavier.org>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.core.models.sqlite;

import androidx.annotation.*;

import org.isoron.uhabits.core.database.*;
import org.isoron.uhabits.core.models.*;
import org.isoron.uhabits.core.models.memory.*;
import org.isoron.uhabits.core.models.sqlite.records.*;
import org.jetbrains.annotations.NotNull;

import java.util.*;

import javax.inject.*;

/**
 * Implementation of a {@link HabitList} that is backed by SQLite.
 */
public class SQLiteHabitList extends HabitList
{
    @NonNull
    private final Repository<HabitRecord> repository;

    @NonNull
    private final ModelFactory modelFactory;

    @NonNull
    private final MemoryHabitList list;

    private boolean loaded = false;

    @Inject
    public SQLiteHabitList(@NonNull ModelFactory modelFactory)
    {
        super();
        this.modelFactory = modelFactory;
        this.list = new MemoryHabitList();
        this.repository = modelFactory.buildHabitListRepository();
    }

    private void loadRecords()
    {
        if (loaded) return;
        loaded = true;

        list.removeAll();
        List<HabitRecord> records = repository.findAll("order by position");

        int expectedPosition = 0;
        boolean shouldRebuildOrder = false;
        for (HabitRecord rec : records)
        {
            if (rec.position != expectedPosition) shouldRebuildOrder = true;
            expectedPosition++;

            Habit h = modelFactory.buildHabit();
            rec.copyTo(h);
            ((SQLiteEntryList) h.getOriginalEntries()).setHabitId(h.getId());
            list.add(h);
        }

        if(shouldRebuildOrder) rebuildOrder();
    }

    @Override
    public synchronized void add(@NonNull Habit habit)
    {
        loadRecords();
        habit.setPosition(size());

        HabitRecord record = new HabitRecord();
        record.copyFrom(habit);
        repository.save(record);
        habit.setId(record.id);
        ((SQLiteEntryList) habit.getOriginalEntries()).setHabitId(record.id);

        list.add(habit);
        getObservable().notifyListeners();
    }

    @Override
    @Nullable
    public synchronized Habit getById(long id)
    {
        loadRecords();
        return list.getById(id);
    }

    @Override
    @Nullable
    public synchronized Habit getByUUID(String uuid)
    {
        loadRecords();
        return list.getByUUID(uuid);
    }

    @Override
    @NonNull
    public synchronized Habit getByPosition(int position)
    {
        loadRecords();
        return list.getByPosition(position);
    }

    @NonNull
    @Override
    public synchronized HabitList getFiltered(HabitMatcher filter)
    {
        loadRecords();
        return list.getFiltered(filter);
    }

    @Override
    @NonNull
    public Order getPrimaryOrder()
    {
        return list.getPrimaryOrder();
    }

    @Override
    public Order getSecondaryOrder()
    {
        return list.getSecondaryOrder();
    }

    @Override
    public synchronized void setPrimaryOrder(@NonNull Order order)
    {
        list.setPrimaryOrder(order);
        getObservable().notifyListeners();
    }

    @Override
    public synchronized void setSecondaryOrder(@NonNull Order order)
    {
        list.setSecondaryOrder(order);
        getObservable().notifyListeners();
    }

    @Override
    public synchronized int indexOf(@NonNull Habit h)
    {
        loadRecords();
        return list.indexOf(h);
    }

    @NotNull
    @Override
    public synchronized Iterator<Habit> iterator()
    {
        loadRecords();
        return list.iterator();
    }

    private synchronized void rebuildOrder()
    {
        List<HabitRecord> records = repository.findAll("order by position");
        repository.executeAsTransaction(() ->
        {
            int pos = 0;
            for (HabitRecord r : records)
            {
                if (r.position != pos)
                {
                    r.position = pos;
                    repository.save(r);
                }
                pos++;
            }
        });
    }

    @Override
    public synchronized void remove(@NonNull Habit habit)
    {
        loadRecords();

        reorder(habit, list.getByPosition(size() - 1));

        list.remove(habit);

        HabitRecord record = repository.find(habit.getId());
        if (record == null) throw new RuntimeException("habit not in database");
        repository.executeAsTransaction(() ->
        {
            habit.getOriginalEntries().clear();
            repository.remove(record);
        });

        getObservable().notifyListeners();
    }

    @Override
    public synchronized void removeAll()
    {
        list.removeAll();
        repository.execSQL("delete from habits");
        repository.execSQL("delete from repetitions");
        getObservable().notifyListeners();
    }

    @Override
    public synchronized void reorder(@NonNull Habit from, @NonNull Habit to)
    {
        loadRecords();
        list.reorder(from, to);

        HabitRecord fromRecord = repository.find(from.getId());
        HabitRecord toRecord = repository.find(to.getId());

        if (fromRecord == null)
            throw new RuntimeException("habit not in database");
        if (toRecord == null)
            throw new RuntimeException("habit not in database");

        if (toRecord.position < fromRecord.position)
        {
            repository.execSQL("update habits set position = position + 1 " +
                               "where position >= ? and position < ?",
                                toRecord.position, fromRecord.position);
        }
        else
        {
            repository.execSQL("update habits set position = position - 1 " +
                               "where position > ? and position <= ?",
                               fromRecord.position, toRecord.position);
        }

        fromRecord.position = toRecord.position;
        repository.save(fromRecord);

        getObservable().notifyListeners();
    }

    @Override
    public synchronized void repair()
    {
        loadRecords();
        rebuildOrder();
        getObservable().notifyListeners();
    }

    @Override
    public synchronized int size()
    {
        loadRecords();
        return list.size();
    }

    @Override
    public synchronized void update(List<Habit> habits)
    {
        loadRecords();
        list.update(habits);

        for (Habit h : habits)
        {
            HabitRecord record = repository.find(h.getId());
            if (record == null) continue;
            record.copyFrom(h);
            repository.save(record);
        }

        getObservable().notifyListeners();
    }

    @Override
    public void resort()
    {
        list.resort();
        getObservable().notifyListeners();
    }

    public synchronized void reload()
    {
        loaded = false;
    }
}

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/models/sqlite/SQLiteHabitList.kt
/*
 * Copyright (C) 2016-2021 Álinson Santos Xavier <git@axavier.org>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package org.isoron.uhabits.core.models.sqlite

import org.isoron.uhabits.core.database.Repository
import org.isoron.uhabits.core.models.Habit
import org.isoron.uhabits.core.models.HabitList
import org.isoron.uhabits.core.models.HabitMatcher
import org.isoron.uhabits.core.models.ModelFactory
import org.isoron.uhabits.core.models.memory.MemoryHabitList
import org.isoron.uhabits.core.models.sqlite.records.HabitRecord
import javax.inject.Inject

/**
 * Implementation of a [HabitList] that is backed by SQLite.
 */
class SQLiteHabitList @Inject constructor(private val modelFactory: ModelFactory) : HabitList() {
    private val repository: Repository<HabitRecord> = modelFactory.buildHabitListRepository()
    private val list: MemoryHabitList = MemoryHabitList()
    private var loaded = false
    private fun loadRecords() {
        if (loaded) return
        loaded = true
        list.groupId = this.groupId
        list.removeAll()
        val records = repository.findAll("order by group_id, position")
        var shouldRebuildOrder = false
        var currentGroup: Long? = null
        var expectedPosition = 0
        for (rec in records) {
            if (currentGroup != rec.groupId) {
                currentGroup = rec.groupId
                expectedPosition = 0
            }
            if (rec.position != expectedPosition) shouldRebuildOrder = true
            val h = modelFactory.buildHabit()
            rec.copyTo(h)
            (h.originalEntries as SQLiteEntryList).habitId = h.id
            if (h.groupId == list.groupId) list.add(h)
            expectedPosition++
        }
        if (shouldRebuildOrder) rebuildOrder()
    }

    @Synchronized
    override fun add(habit: Habit) {
        loadRecords()
        habit.position = size()
        habit.id = repository.getNextAvailableId("habitandgroup")
        val record = HabitRecord()
        record.copyFrom(habit)
        repository.save(record)
        (habit.originalEntries as SQLiteEntryList).habitId = record.id
        list.add(habit)
        observable.notifyListeners()
    }

    @Synchronized
    private fun rebuildOrder() {
        val records = repository.findAll("order by group_id, position")
        repository.executeAsTransaction {
            var currentGroup: Long? = null
            var expectedPosition = 0
            for (r in records) {
                if (currentGroup != r.groupId) {
                    currentGroup = r.groupId
                    expectedPosition = 0
                }
                if (r.position != expectedPosition) {
                    r.position = expectedPosition
                    repository.save(r)
                }
                expectedPosition++
            }
        }
    }

    @Synchronized
    override fun add(position: Int, habit: Habit) {
        loadRecords()
        habit.position = size()
        val record = HabitRecord()
        record.copyFrom(habit)
        repository.save(record)
        habit.id = record.id
        (habit.originalEntries as SQLiteEntryList).habitId = record.id
        list.add(position, habit)
        observable.notifyListeners()
    }

    @Synchronized
    override fun getById(id: Long): Habit? {
        loadRecords()
        return list.getById(id)
    }

    @Synchronized
    override fun getByUUID(uuid: String?): Habit? {
        loadRecords()
        return list.getByUUID(uuid)
    }

    @Synchronized
    override fun getByPosition(position: Int): Habit {
        loadRecords()
        return list.getByPosition(position)
    }

    @Synchronized
    override fun getFiltered(matcher: HabitMatcher?): HabitList {
        loadRecords()
        return list.getFiltered(matcher)
    }

    @set:Synchronized
    override var primaryOrder: Order
        get() = list.primaryOrder
        set(order) {
            list.primaryOrder = order
            observable.notifyListeners()
        }

    @set:Synchronized
    override var secondaryOrder: Order
        get() = list.secondaryOrder
        set(order) {
            list.secondaryOrder = order
            observable.notifyListeners()
        }

    @Synchronized
    override fun indexOf(h: Habit): Int {
        loadRecords()
        return list.indexOf(h)
    }

    @Synchronized
    override fun iterator(): Iterator<Habit> {
        loadRecords()
        return list.iterator()
    }

    @Synchronized
    override fun remove(h: Habit) {
        loadRecords()
        list.remove(h)
        val record = repository.find(
            h.id!!
        ) ?: throw RuntimeException("habit not in database")
        repository.executeAsTransaction {
            h.originalEntries.clear()
            repository.remove(record)
        }
        rebuildOrder()
        observable.notifyListeners()
    }

    @Synchronized
    override fun removeAt(position: Int) {
        loadRecords()
        list.removeAt(position)
    }

    @Synchronized
    override fun removeAll() {
        list.removeAll()
        repository.execSQL("delete from habits")
        repository.execSQL("delete from repetitions")
        observable.notifyListeners()
    }

    @Synchronized
    override fun reorder(from: Habit, to: Habit) {
        loadRecords()
        list.reorder(from, to)
        val fromRecord = repository.find(
            from.id!!
        )
        val toRecord = repository.find(
            to.id!!
        )
        if (fromRecord == null) throw RuntimeException("habit not in database")
        if (toRecord == null) throw RuntimeException("habit not in database")
        if (toRecord.position!! < fromRecord.position!!) {
            repository.execSQL(
                "update habits set position = position + 1 " +
                    "where position >= ? and position < ?",
                toRecord.position!!,
                fromRecord.position!!
            )
        } else {
            repository.execSQL(
                "update habits set position = position - 1 " +
                    "where position > ? and position <= ?",
                fromRecord.position!!,
                toRecord.position!!
            )
        }
        fromRecord.position = toRecord.position
        repository.save(fromRecord)
        observable.notifyListeners()
    }

    @Synchronized
    override fun repair() {
        loadRecords()
        rebuildOrder()
        observable.notifyListeners()
    }

    @Synchronized
    override fun size(): Int {
        loadRecords()
        return list.size()
    }

    @Synchronized
    override fun update(habits: List<Habit>) {
        loadRecords()
        list.update(habits)
        for (h in habits) {
            val record = repository.find(h.id!!) ?: continue
            record.copyFrom(h)
            repository.save(record)
        }
        observable.notifyListeners()
    }

    override fun resort() {
        list.resort()
        observable.notifyListeners()
    }

    @Synchronized
    fun reload() {
        loaded = false
    }
}
